### PR TITLE
Adds a checkbox to select all issues/PRs

### LINKED ISSUE #18883: Way to bulk-select issues in a list
### Feature Description

For example, there could be a checkbox above the first issue. Clicking that box would select all of the issues that are visible. This would be useful when changing the labels or milestone of many issues at once.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by tyroneyeh**
Closes #18883 

Adds a checkbox to select all

image

### CODE REVIEW

**Comment by shugen002** [comment]
https://drone.gitea.io/go-gitea/gitea/56916/1/4 
lint problem.

**Comment by tyroneyeh** [comment]
> https://drone.gitea.io/go-gitea/gitea/56916/1/4 lint problem.

Thanks, Fixed

File: web_src/js/features/common-issue.js [hunk 1]
```diff
@@ -2,14 +2,33 @@ import $ from 'jquery';
 import {updateIssuesMeta} from './repo-issue.js';
 
 export function initCommonIssue() {
-  $('.issue-checkbox').on('click', () => {
-    const numChecked = $('.issue-checkbox').children('input:checked').length;
-    if (numChecked > 0) {
+  $('.issue-checkbox,.issue-checkbox-all').on('click', (e) => {
```
Comment (Gusted) [inline web_src/js/features/common-issue.js]: Should this be in one click event? Seems better for code readability to separate it.

File: web_src/js/features/common-issue.js [hunk 2]
```diff
@@ -2,14 +2,33 @@ import $ from 'jquery';
...
+      }
+    }
+    if (e.shiftKey && window.config.checkboxfirst !== undefined) {
+      for (let i = window.config.checkboxfirst + 1, j = issuecheckbox.index($(e.currentTarget).find('input')); i < j; i++) {
+        issuecheckbox[i].checked = 1;
+      }
+      delete window.config.checkboxfirst;
+    } else {
+      window.config.checkboxfirst = issuecheckbox.index($(e.currentTarget).find('input'));
+    }
```
Comment (Gusted) [inline web_src/js/features/common-issue.js]: `window.config` shouldn't be used to store data. This rather should be a variable declared within this function.

**Comment by tyroneyeh** [comment]
is that so?

File: web_src/js/features/common-issue.js [hunk 3]
```diff
@@ -2,15 +2,36 @@ import $ from 'jquery';
...
     } else {
       $('#issue-filters').removeClass('hide');
       $('#issue-actions').addClass('hide');
+      $('#issue-filters .six').prepend($('.issue-checkbox-all'));
     }
+  }
+
+  $('.issue-checkbox').on('click', (e) => {
+    checkboxOperate(e);
+  });
```
Comment (Gusted) [inline web_src/js/features/common-issue.js]:
```suggestion
  $('.issue-checkbox').on('click', checkboxOperate);
```

File: web_src/js/features/common-issue.js [hunk 4]
```diff
@@ -2,15 +2,36 @@ import $ from 'jquery';
 import {updateIssuesMeta} from './repo-issue.js';
 
 export function initCommonIssue() {
-  $('.issue-checkbox').on('click', () => {
-    const numChecked = $('.issue-checkbox').children('input:checked').length;
-    if (numChecked > 0) {
+  function checkboxOperate(e) {
```
Comment (Gusted) [inline web_src/js/features/common-issue.js]:
```suggestion
  const checkboxOperate = (e) => {
```

File: web_src/js/features/common-issue.js [hunk 5]
```diff
@@ -2,15 +2,34 @@ import $ from 'jquery';
...
-  $('.issue-checkbox').on('click', () => {
-    const numChecked = $('.issue-checkbox').children('input:checked').length;
-    if (numChecked > 0) {
+  const checkboxOperate = (e) => {
+    const issuecheckbox = $('.issue-checkbox input');
+    if (e.shiftKey && window.checkboxfirst !== undefined) {
+      for (let i = window.checkboxfirst + 1, j = issuecheckbox.index($(e.currentTarget).find('input')); i < j; i++) {
+        issuecheckbox[i].checked = 1;
+      }
+      delete window.checkboxfirst;
```
Comment (Gusted) [inline web_src/js/features/common-issue.js]: Please use a local variable for this, inside the `initCommonIssue` function.

Comment (tyroneyeh) [reply web_src/js/features/common-issue.js]: Is that so?
https://github.com/go-gitea/gitea/pull/20177/commits/25fa1d85920d668ce07c3f8be2cac796a50ddd95

Comment (Gusted) [reply web_src/js/features/common-issue.js]: That's better. No need to use `window.`

**Review by Gusted — APPROVED** [review]

**Comment by tyroneyeh** [comment]
PTAL

File: templates/repo/issue/list.tmpl [hunk 6]
```diff
@@ -28,6 +28,12 @@
 		<div class="ui divider"></div>
 		<div id="issue-filters" class="ui stackable grid">
 			<div class="six wide column">
+				{{if $.CanWriteIssuesOrPulls}}
+				<div class="ui checkbox issue-checkbox-all">
+					<input type="checkbox"></input>
+					<label></label>
+				</div>
```
Comment (Gusted) [inline templates/repo/issue/list.tmpl]:
```suggestion
					<div class="ui checkbox issue-checkbox-all">
						<input type="checkbox"></input>
						<label></label>
					</div>
```

File: web_src/js/features/common-issue.js [hunk 7]
```diff
@@ -2,15 +2,34 @@ import $ from 'jquery';
 import {updateIssuesMeta} from './repo-issue.js';
 
 export function initCommonIssue() {
-  $('.issue-checkbox').on('click', () => {
-    const numChecked = $('.issue-checkbox').children('input:checked').length;
-    if (numChecked > 0) {
+  let checkboxfirst;
+  const checkboxOperate = (e) => {
+    const issuecheckbox = $('.issue-checkbox input');
+    if (e.shiftKey && checkboxfirst !== undefined) {
```
Comment (Gusted) [inline web_src/js/features/common-issue.js]: This currently only works when the checkbox is below the previous checkbox. Which can be confusing behavior that the other way around doesn't work, also might be something to do for another PR to minimize the changes in this PR.

**Review by Gusted — COMMENTED** [review]
Otherwise L-GTM

File: web_src/js/features/common-issue.js (hunk 7 above)
Comment (tyroneyeh) [reply web_src/js/features/common-issue.js]: Thanks, ok, i deleted it

**Review by Gusted — APPROVED** [review]

File: templates/repo/issue/list.tmpl [hunk 8]
```diff
@@ -28,6 +28,12 @@
 		<div class="ui divider"></div>
 		<div id="issue-filters" class="ui stackable grid">
 			<div class="six wide column">
+				{{if $.CanWriteIssuesOrPulls}}
+					<div class="ui checkbox issue-checkbox-all">
+						<input type="checkbox"></input>
+						<label></label>
```
Comment (delvh) [inline templates/repo/issue/list.tmpl:34]: Should there perhaps be a small text like `All`?

**Comment by Gusted** [comment]
> Merge main failed?

Flaky tests. Nothing to worry about.

File: web_src/js/features/common-issue.js [hunk 9]
```diff
@@ -2,15 +2,25 @@ import $ from 'jquery';
...
-    if (numChecked > 0) {
+  const checkboxOperate = () => {
+    if ($('.issue-checkbox input').is(':checked')) {
       $('#issue-filters').addClass('hide');
       $('#issue-actions').removeClass('hide');
+      $('#issue-actions .six').prepend($('.issue-checkbox-all'));
     } else {
       $('#issue-filters').removeClass('hide');
       $('#issue-actions').addClass('hide');
+      $('#issue-filters .six').prepend($('.issue-checkbox-all'));
```
Comment (wxiaoguang) [inline web_src/js/features/common-issue.js]: What's the purpose of these two `prepend($('.issue-checkbox-all'));`?

File: web_src/js/features/common-issue.js [hunk 10]
```diff
@@ -2,15 +2,25 @@ import $ from 'jquery';
...
     }
+  };
+
+  $('.issue-checkbox').on('click', checkboxOperate);
+
+  $('.issue-checkbox-all').on('click', (e) => {
+    const selected = $('.issue-checkbox input:checked');
+    $('.issue-checkbox input:not(:checked)').prop('checked', 1);
+    selected.prop('checked', 0);
+    checkboxOperate(e);
```
Comment (wxiaoguang) [inline web_src/js/features/common-issue.js]: If I understand correctly, the code does invert selection while the name is `issue-checkbox-all` (sounds like check all), it's a little misleading.

And `checkboxOperate` could have a better and clearer name.

File: web_src/js/features/common-issue.js (hunk 9 above)
Comment (tyroneyeh) [reply web_src/js/features/common-issue.js]: If the switch is checked, move the Check All checkbox in the template to the issue filter or action area

File: web_src/js/features/common-issue.js (hunk 10 above)
Comment (tyroneyeh) [reply web_src/js/features/common-issue.js]: This can be clicked on the child checkbox in the checked all checkbox can invert the selection of the child
Any suggestions on function names?

File: web_src/js/features/common-issue.js (hunk 9 above)
Comment (wxiaoguang) [reply web_src/js/features/common-issue.js]: Just tried, the layout here was quite interesting before. It's right!

File: web_src/js/features/common-issue.js (hunk 10 above)
Comment (wxiaoguang) [reply web_src/js/features/common-issue.js]: I think it can be called `issue-checkbox-invert`.

I will add some comments here and approve & merge.

Comment (tyroneyeh) [reply web_src/js/features/common-issue.js]: `issue-checkbox-invert` this is function name? or `issueCheckboxInvert` ?

Comment (wxiaoguang) [reply web_src/js/features/common-issue.js]: After some testing, I feel that the `all/invert` behavior is still a little confusing. For example, we can get a UI like below.

I do not think these are common UIs for end users, the `all/invert` checkbox's state is confusing. Maybe the `all/invert` element should be replaced by a icon button, without state.

<details>

image

image


image

image

</details>

Comment (wxiaoguang) [reply web_src/js/features/common-issue.js]: To make UI consistent, I think there could be 2 choices:

* Make the `checkbox-all` really do check all/check none
* Use a `checkbox-invert` button (without state) to toggle/invert selection

Personally I think choice 1 (check all/check none) is clear and still easy to use. How do you think?

Comment (tyroneyeh) [reply web_src/js/features/common-issue.js]: You mean don't invert selection of parts, just select all and deselect all?

Comment (wxiaoguang) [reply web_src/js/features/common-issue.js]: yes

Comment (tyroneyeh) [reply web_src/js/features/common-issue.js]: Is it like this? https://github.com/go-gitea/gitea/pull/20177/commits/01f3f0f14d5a45fee9f1452e9469a955b1efecf2

Comment (wxiaoguang) [reply web_src/js/features/common-issue.js]: Thank you ~~ I also do some improvments. Now the checkbox states are synchronized and the select-all checkbox has an indeterminate state.

image

**Review by wxiaoguang — APPROVED** [review]

**Review by delvh — APPROVED** [review]

File: templates/repo/issue/list.tmpl (hunk 8 above)
Comment (wxiaoguang) [reply templates/repo/issue/list.tmpl:34]: IMO it's convention that the checkbox in the first column is for select-all, so maybe it's not a must to add texts here.

There are a lot of UI for reference. Google "select all checkbox ui" for images. 😁 

----

And here is the Gmail UI:

image

Comment (delvh) [reply templates/repo/issue/list.tmpl:34]: From what I've seen, there seems to be an equilibrium between applications that do it and those that don't.
I don't care which one we use, as long as we decide on one.

**Comment by wxiaoguang** [comment]
bot
